subsys/shell/backends/shell_mqtt.c

```diff
@@ -0,0 +1,824 @@
+
+SHELL_MQTT_DEFINE(shell_transport_mqtt);
+SHELL_DEFINE(shell_mqtt, "", &shell_transport_mqtt,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_SIZE,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_TIMEOUT, SHELL_FLAG_OLF_CRLF);
+
+LOG_MODULE_REGISTER(shell_mqtt, CONFIG_SHELL_MQTT_LOG_LEVEL);
+
+#define NET_EVENT_MASK (NET_EVENT_L4_CONNECTED | NET_EVENT_L4_DISCONNECTED)
+#define CONNECT_TIMEOUT_MS 2000
+#define LISTEN_TIMEOUT_MS 500
+#define MQTT_SEND_DELAY_MS K_MSEC(100)
+#define PROCESS_INTERVAL K_SECONDS(2)
+#define SHELL_MQTT_WORKQ_STACK_SIZE 2048
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_USERNAME
+#define MQTT_USERNAME CONFIG_SHELL_MQTT_SERVER_USERNAME
+#else
+#define MQTT_USERNAME NULL
+#endif /* CONFIG_SHELL_MQTT_SERVER_USERNAME */
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#define MQTT_PASSWORD CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#else
+#define MQTT_PASSWORD NULL
+#endif /*SHELL_MQTT_SERVER_PASSWORD */
+
+struct shell_mqtt *sh_mqtt;
+K_KERNEL_STACK_DEFINE(sh_mqtt_workq_stack, SHELL_MQTT_WORKQ_STACK_SIZE);
+
+static void mqtt_evt_handler(struct mqtt_client *const client, const struct mqtt_evt *evt);
+
+static inline int sh_mqtt_work_reschedule(struct k_work_delayable *dwork, k_timeout_t delay)
+{
+	return k_work_reschedule_for_queue(&sh_mqtt->workq, dwork, delay);
+}
+
+static inline int sh_mqtt_work_submit(struct k_work *work)
+{
+	return k_work_submit_to_queue(&sh_mqtt->workq, work);
+}
+
+/* Lock the context of the shell mqtt */
+static inline int sh_mqtt_context_lock(k_timeout_t timeout)
+{
+	return k_mutex_lock(&sh_mqtt->lock, timeout);
+}
+
+/* Unlock the context of the shell mqtt */
+static inline void sh_mqtt_context_unlock(void)
+{
+	k_mutex_unlock(&sh_mqtt->lock);
+}
+
+bool __weak shell_mqtt_get_devid(char *id, int id_max_len)
+{
+	uint8_t hwinfo_id[DEVICE_ID_BIN_MAX_SIZE];
+	ssize_t length;
+
+	length = hwinfo_get_device_id(hwinfo_id, DEVICE_ID_BIN_MAX_SIZE);
+	if (length <= 0) {
+		return false;
+	}
+
+	(void)memset(id, 0, id_max_len);
+	length = bin2hex(hwinfo_id, (size_t)length, id, id_max_len - 1);
+
+	return length > 0;
+}
+
+static void prepare_fds(void)
+{
+	if (sh_mqtt->mqtt_cli.transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		sh_mqtt->fds[0].fd = sh_mqtt->mqtt_cli.transport.tcp.sock;
+	}
+
+	sh_mqtt->fds[0].events = ZSOCK_POLLIN;
+	sh_mqtt->nfds = 1;
+}
+
+static void clear_fds(void)
+{
+	sh_mqtt->nfds = 0;
+}
+
+static int wait(int timeout)
+{
+	int rc = 0;
+
+	if (sh_mqtt->nfds > 0) {
+		rc = zsock_poll(sh_mqtt->fds, sh_mqtt->nfds, timeout);
+		if (rc < 0) {
+			LOG_ERR("poll error: %d", errno);
+		}
+	}
+
+	return rc;
+}
+
+/* Query IP address for the broker URL */
+static int get_mqtt_broker_addrinfo(void)
+{
+	int rc;
+
+	sh_mqtt->hints.ai_family = AF_INET;
+	sh_mqtt->hints.ai_socktype = SOCK_STREAM;
+	sh_mqtt->hints.ai_protocol = 0;
+
+	rc = zsock_getaddrinfo(CONFIG_SHELL_MQTT_SERVER_ADDR,
+			       STRINGIFY(CONFIG_SHELL_MQTT_SERVER_PORT), &sh_mqtt->hints,
+			       &sh_mqtt->haddr);
```

Comment:
Note that `zsock_getaddrinfo` will allocate memory for the resulting address - if it's not freed at some point, we end up with a memory leak. I saw this function is called from the connect handler, therefore it could be called multiple times during runtime? I think it should either be called once at init, if we want to skip memory freeing, or perhaps call `zsock_freeaddrinfo()` if `sh_mqtt->haddr != NULL` before calling `zsock_getaddrinfo()`.

Reply:
Fixed, thanks